Zend/zend_alloc.c

```diff
@@ -296,6 +296,20 @@ static bool zend_mm_use_huge_pages = false;
  */
 
 struct _zend_mm_heap {
+    union {                                             /* This union contains security-relevant properties, and is thus made read-only at run-time. */
+        struct {
+            uintptr_t          shadow_key;              /* free slot shadow ptr xor key */
+#if ZEND_MM_CUSTOM
+            struct {
+                void      *(*_malloc)(size_t ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void       (*_free)(void*  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void      *(*_realloc)(void*, size_t  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+            } custom_heap;
+#endif
+        };
+        char padding [ZEND_MM_PAGE_SIZE];
+    };
+
 #if ZEND_MM_CUSTOM
 	int                use_custom_heap;
 #endif
```

Comment:
We are toggling this flag in the profiler before we forward `malloc`/`free`/`realloc` calls back to the original heap. Currently this is the first element of the `zend_mm_heap`, so it is easy for us to toggle this. By moving this below your union it will get les stable to calculate the address of this flag.

Reply:
I am currently checking if https://github.com/php/php-src/pull/13432 is sufficient to not rely on the "tinkering with `zend_mm_heap->use_custom_heap` flag"-hack anymore

Reply:
Looking good so far 😄